src/Kiota.Builder/KiotaBuilder.cs

```diff
@@ -1548,43 +1548,56 @@ private CodeType CreateModelDeclarationAndType(OpenApiUrlTreeNode currentNode, O
+    {
+        var flattenedAllOfs = schema.AllOf.FlattenSchemaIfRequired(static x => x.AllOf).ToArray();
+        var referenceId = schema.Reference?.Id;
+        var className = (schema.GetSchemaName(true) is string cName && !string.IsNullOrEmpty(cName) ?
                             cName :
-                            currentNode.GetClassName(config.StructuredMimeTypes, operation: operation, suffix: classNameSuffix, schema: schema, requestBody: isRequestBody))
+                            (!string.IsNullOrEmpty(typeNameForInlineSchema) ?
+                                typeNameForInlineSchema :
+                                currentNode.GetClassName(config.StructuredMimeTypes, operation: operation, suffix: classNameSuffix, schema: schema, requestBody: isRequestBody)))
                         .CleanupSymbolName();
-            if (shortestNamespace != null)
-                codeDeclaration = AddModelDeclarationIfDoesntExist(currentNode, currentSchema, className, shortestNamespace, codeDeclaration as CodeClass);
-        }
-        if (codeDeclaration is CodeClass currentClass &&
-            !currentClass.Documentation.DescriptionAvailable &&
+        var shortestNamespaceName = GetModelsNamespaceNameFromReferenceId(referenceId);
+        var codeNamespaceFromParent = GetShortestNamespace(codeNamespace, schema);
+        if (rootNamespace is null)
+            throw new InvalidOperationException("Root namespace is not set");
+        var shortestNamespace = string.IsNullOrEmpty(referenceId) ? codeNamespaceFromParent : rootNamespace.FindOrAddNamespace(shortestNamespaceName);
+        var inlineSchema = Array.Find(flattenedAllOfs, static x => !x.IsReferencedSchema());
+        var referencedSchema = Array.Find(flattenedAllOfs, static x => x.IsReferencedSchema());
+        var rootSchemaHasProperties = schema.HasAnyProperty();
+        var codeDeclaration = (rootSchemaHasProperties, inlineSchema, referencedSchema) switch
+        {
+            // greatest parent type
+            (true, null, null) => AddModelDeclarationIfDoesntExist(currentNode, schema, className, shortestNamespace),
+            // inline schema + referenced schema
+            (false, not null, not null) => AddModelDeclarationIfDoesntExist(currentNode, inlineSchema, className, shortestNamespace, CreateInheritedModelDeclaration(currentNode, referencedSchema, operation, classNameSuffix, codeNamespace, isRequestBody, typeNameForInlineSchema)),
+            // properties + referenced schema
+            (true, null, not null) => AddModelDeclarationIfDoesntExist(currentNode, schema, className, shortestNamespace, CreateInheritedModelDeclaration(currentNode, referencedSchema, operation, classNameSuffix, codeNamespace, isRequestBody, typeNameForInlineSchema)),
```

Comment:
```suggestion
            (true, null, not null) => AddModelDeclarationIfDoesntExist(currentNode, schema, className, shortestNamespace, CreateInheritedModelDeclaration(currentNode, referencedSchema, operation, classNameSuffix, codeNamespace, isRequestBody, string.Empty)),
```
Same reasoning. Since we already know we are trying to create a model from an reference schema there is really no need to pass forward the info. It may lead to creating a model with a name we don't want or be unable to resolve duplicates.